### PR TITLE
Update cellular test environment setup and test cases

### PR DESCRIPTION
**Pull request by chinglee-iot**
Update cellular test environment setup and test cases

Description
-----------

* Add BOARD_HAS_CELLULAR compile option to enable cellular support
  - To use cellular HAL API, define BOARD_HAS_CELLULAR=1
  - To use secure socket cellular, define BOARD_HAS_CELLULAR=1 and
   SECURE_SOCKETS_CELLULAR=1

* Setup cellular connection in test application

* Add new cellular test cases
  - AFQP_Cellular_GetNetworkTime
  - AFQP_Cellular_GetHostByName
  - AFQP_Cellular_AtCommandRawAndGenericUrc
  - AFQP_Cellular_EidrxEchoTimes
  - AFQP_Cellular_PsmStatus

* Update cellular connection setup flow

* Fix windows simulator comm interface hang problem

* Temporarily disable UART common I/O when set BOARD_HAS_CELLULAR=1 in
  st_discovery board

Checklist:
----------

- [ ] I have tested my changes. No regression in existing tests.
- [ ] My code is Linted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: vendors/st/boards/stm32l475_discovery/aws_tests/application_code/main.c [hunk 1]
```diff
@@ -109,6 +112,17 @@ int main( void )
 
     return 0;
 }
+
+/*-----------------------------------------------------------*/
+
+static void testRunnerHook( void * pvParameters )
```
Comment (aggarg) [inline vendors/st/boards/stm32l475_discovery/aws_tests/application_code/main.c:118]: testRunnerTask is probably better?

File: vendors/st/boards/stm32l475_discovery/aws_tests/application_code/main.c [hunk 2]
```diff
@@ -66,6 +66,9 @@ void vApplicationDaemonTaskStartupHook( void );
 extern void SPI_WIFI_ISR(void);
 extern SPI_HandleTypeDef hspi;
 
+/* Setup cellular connection. */
+bool prvSetupCellular( void );
```
Comment (aggarg) [inline vendors/st/boards/stm32l475_discovery/aws_tests/application_code/main.c:70]: Is this a private (i.e. file static) function?  If not, it should not be named prv. Also, does the call site need to check return value?

File: vendors/st/boards/stm32l475_discovery/aws_tests/STM32L475VGTx_FLASH.ld [hunk 3]
```diff
@@ -43,8 +43,8 @@ MEMORY
 {
 RAM (xrw)      : ORIGIN = 0x20000000, LENGTH = 96K
 RAM2 (xrw)      : ORIGIN = 0x10000000, LENGTH = 32K
-FLASH (rx)      : ORIGIN = 0x8000000, LENGTH = 400K		/* Use only the first bank */
-FLASH_UC (r)	: ORIGIN = 0x08064000, LENGTH = 9K		/* Fixed-location area */
+FLASH (rx)      : ORIGIN = 0x8000000, LENGTH = 512K		/* Use only the first bank */
```
Comment (aggarg) [inline vendors/st/boards/stm32l475_discovery/aws_tests/STM32L475VGTx_FLASH.ld:46]: Why is this change needed? The comment says, use only first bank and if  I remember correctly, size of each bank is 512K.

File: libraries/abstractions/cellular/test/test_cellular_api.c [hunk 4]
```diff
@@ -97,88 +149,77 @@
...
-
-    configPRINTF( ( "AT command set and ready. \r\n" ) );
-    return CELLULAR_SUCCESS;
-}
+/*-----------------------------------------------------------*/
 
-static void prvsignalStrengthChangedCallback( CellularUrcEvent_t urcEvent,
-                                              const CellularSignalInfo_t * pSignalInfo )
+/* Signal strength changed callback function. */
+static void prvSignalStrengthChangedCallbac( CellularUrcEvent_t urcEvent,
```
Comment (aggarg) [inline libraries/abstractions/cellular/test/test_cellular_api.c:221]: prvSignalStrengthChangedCallbac**k**

Comment (chinglee-iot) [reply libraries/abstractions/cellular/test/test_cellular_api.c:221]: This will be fixed.

File: vendors/st/boards/stm32l475_discovery/aws_tests/application_code/main.c (hunk 1 above)
Comment (chinglee-iot) [reply vendors/st/boards/stm32l475_discovery/aws_tests/application_code/main.c:118]: The name of the function will be updated. Thanks for the suggestion.

File: vendors/st/boards/stm32l475_discovery/aws_tests/application_code/main.c (hunk 2 above)
Comment (chinglee-iot) [reply vendors/st/boards/stm32l475_discovery/aws_tests/application_code/main.c:70]: Originally, it is designed as a private function. It should only be used in the main.c file.
To reduce the modification in main.c, the implementation of prvSetupCellular is moved to cellular_setup.c.

The return value should be checked. A message will be added if this function returns fail.

File: vendors/st/boards/stm32l475_discovery/aws_tests/STM32L475VGTx_FLASH.ld (hunk 3 above)
Comment (chinglee-iot) [reply vendors/st/boards/stm32l475_discovery/aws_tests/STM32L475VGTx_FLASH.ld:46]: I think 400K is enough for test application. This will be revert back.

Currently footprint for cellular
Cellular common : 41K
MbedTLS : 60K
Secure sockets cellular : 11K

Also the FLASH_UC is overlapped. It will be updated.
aws_demos
FLASH (rx)      : ORIGIN = 0x08000000, LENGTH = 503K    /* Use only the first bank */
FLASH_UC (r)	: ORIGIN = 0x0807DC00, LENGTH = 9K		/* Fixed-location area */

aws_tests:
FLASH (rx)      : ORIGIN = 0x08000000, LENGTH = 400K    /* Use only the first bank */
FLASH_UC (r)	: ORIGIN = 0x08064000, LENGTH = 9K		/* Fixed-location area */

**Review by aggarg — APPROVED** [review]
